### ISSUE TITLE
Orders without equivalence relations

### ISSUE THREAD

**Issue by laMudri** [agda/agda-stdlib#2070]
Sometimes we come across types which naturally have an order, but for which the appropriate equivalence relation is, intensionally, the intersection of the order and its inverse. For example, if we have types about which we only care about inhabitedness (propositions), then the order is given by `→`, while the equivalence relation is given by bi-implication. Giving this definition using `Poset` from `Relation.Binary` is annoying because we have to prove reflexivity and transitivity for implication, then prove them twice more for equivalence, and then give (simple, but slightly tedious) proofs of symmetry and antisymmetry.

It would be nice if there were some general code to take a type with a reflexive transitive relation and produce the corresponding poset. Could the structure be slotted into the `Relation.Binary` hierarchy, perhaps even before `Setoid`, together with the construction suggested above? I've occasionally heard the term “proset” used for a partially ordered set without an equivalence relation, though nLab doesn't distinguish this from partially ordered sets in the normal sense.

**Comment by gallais** [comment]
I suppose we could have something like:

http://agda.github.io/agda-stdlib/Relation.Binary.Construct.Closure.Equivalence.html

**Comment by MatthewDaggitt** [comment]
Yup adding a `Construct` module seems to be the right way to go about it.

**Comment by laMudri** [comment]
See #2071.

**Comment by jamesmckinna** [comment]
'Inverse' or 'converse' relation? 
One reason I'm less happy on the PR with 'core groupoid' as justification for the terminology 'core' is that you're not actually inverting anything here, are you, in the sense of having left- and right- inverses to an arrow? (Or are you, if I squint hard enough?)

Calling the opposite/converse of a relation 'inverse' seems... perverse, to me at least.

**Comment by jamesmckinna** [comment]
As for the lexicon shift from `core` to `interior` (as a dual to `closure`, in the context of topology originally, AFAIK): 

Just as the `P`-closure of a relation `R` for some property `P` is the *least* relation `S` containing `R` such that `P S`, the `P`-*interior* is the *dual* notion, vs. the *greatest* relation `S` *contained in* `R` such that `P S`, for any suitable `P`. 

This issue/PR addresses the case `P = Symmetric`. 

For a citation, and some sufficient conditions on `P` to be 'suitable', you could look at the treatment of 'intersection-closed' and 'union-closed' families of sets/predicates/relations defined in [Paul M. Cohn's "Universal Algebra" (1965)](https://link.springer.com/book/10.1007/978-94-009-8399-1) and their application to defining closures and interiors, and inductive/coinductive definitions generally (see also Peter Aczel's article on inductive definitions in the Handbook of Mathematical Logic (1977) (ISBN: 9780720422856)). NB impredicative definitions! (cf. Tarski's proof of his fixed-point theorem)

Being `Symmetric` is, IIRC, an example of a property which is both intersection- and union-closed, and hence has a corresponding closure and interior, as here. 

Citation update: the earliest precursor to Cohn that I can find for intersection and union closed families would be Graetzer's seminal text, but I personally find that pretty hard going, not least notationally, but also for the sense of an emerging lexicon for describing the mathematical research domain at hand which is... not quite ready for ordinary mortals to grok, but YMMV.

**Comment by jamesmckinna** [comment]
Regarding the introduction (and subsequent removal;-)) of `Proset` from the `Relation.Bundle` hierarchy, suggest that any such proposal file under the `hypothetical-rewrite` milestone. I think there's potential for it, but as @MatthewDaggitt points out on #2071 such a thing would involve a (much!) large(r)-scale reconsideration of many things, so, for good or bad, IMHO should wait for such time as we have time/person-power to undertake that...

**Comment by jamesmckinna** [comment]
Has this issue been successfully closed by #2071 ?

**Comment by JacquesCarette** [comment]
It doesn't look like #2071 closes this, it 'only' provides the underlying machinery to get there. It feels like this new machinery should get actively used in combinators to make one's life easier, as per the original issue.

**Comment by jamesmckinna** [comment]
Underlying machinery, or the relevant construction?

https://github.com/agda/agda-stdlib/blob/54f5c38799f6c0125f7dc914e073670b549bcbe6/src/Relation/Binary/Construct/Interior/Symmetric.agda#L84-L107

I think some of the tension in recent PRs #2809 / #2721 concerning (D)CPOs turns on this?

Or did you mean the concrete instance given by biequivalence induced by implication?

**Comment by JacquesCarette** [comment]
@jamesmckinna we're failing to communicate. I think we both have implicit context in our minds that isn't in our messages, and we're thus not achieving communication. I read your latest post multiple times, and I can't make sense of it... Thus my guess. As I'm quite convinced that what you're asking does make sense, but I'm not understanding the flow of the whole conversation. Might need to back up a few steps and restart.

**Comment by jamesmckinna** [comment]
@JacquesCarette Yes, old issues and comments on them can be hard to page back in, when the discussion is over a long period, moreover asynchronous. So, to try to back up, my understanding of the OP was two-fold:
* can we have the construction of a `Poset`/`Preorder`, *and* its underlying `Setoid`, from a bare reflexive, transitive relation, via its symmetric interior... and I believe (as, I hope, do you) that #2071 furnishes such constructions
* that it would be 'nice' if the arrow structure on `Set` could be so lifted so as to build the 'implication preorder', with bi-implication as its symmetric interior. This has, AFAICT, not been done, but should be a downstream PR.

As to closing this issue, for me, it would be/have been clearer if we closed this one, and opened a second, focused one (or simply that sought-after PR) for the implication preorder, but I could see arguments to keep this one open, although with the lag/latency in discussion of it leading to your recent comment.

But the symmetrisation also came up in a comment from @TOTBWF about order theory/domain theory prerequisites, and it was perhaps (my) muddying the water to allude to that here, rather than deal with it there.

If you want to keep this issue open until the implication preorder is developed and merged, fine by me.
If we have a separate issue to treat that, as based on this and #2071, also fine by me, and arguably clearer.

Does this account chime with yours?

**Comment by JacquesCarette** [comment]
Certainly your long comment makes sense. It would make sense to close this issue once a new one about the 'implication preorder' existed as a separate issue.

**Comment by jamesmckinna** [comment]
Done, modulo tripping up over creating a 'sub-issue'. See #2820
